# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: dom/workers/ScriptLoader.cpp
# Commit: 2fd61eb5c69c
# Full Hash: 2fd61eb5c69ce9ac806048a35c7a7a88bf4b9652
# Author: Perry Jiang <perry@mozilla.com>
# Date: 2020-04-21 09:42:20
# Regressor Bug: 1618546
# File Overlap Count: 2
# Description:
#   Bug 1618546 - give worker debugger globals their own clients r=asuth,webidl,smaug
#   
#   - Worker debugger globals gets a client with a null principal
#   - Ensure globals are created before script loads
#   - Introduce WorkerGlobalScopeBase to share code
# ==============================================================================

diff -r 9ccf9f680b64 -r 2fd61eb5c69c dom/workers/ScriptLoader.cpp
--- a/dom/workers/ScriptLoader.cpp	Tue Apr 21 06:09:28 2020 +0000
+++ b/dom/workers/ScriptLoader.cpp	Tue Apr 21 06:50:53 2020 +0000
@@ -1941,7 +1941,7 @@
         // ClientInfo should always be present since this should not be called
         // if parent's status is greater than Running.
         ,
-        mClientInfo(aParentWorker->GetClientInfo().ref()),
+        mClientInfo(aParentWorker->GlobalScope()->GetClientInfo().ref()),
         mLoadInfo(aLoadInfo),
         mResult(NS_ERROR_FAILURE) {
     MOZ_ASSERT(aParentWorker);
@@ -2045,23 +2045,6 @@
   // where the CSP code expects it!
   aWorkerPrivate->StoreCSPOnClient();
 
-  AutoJSAPI jsapi;
-  jsapi.Init();
-
-  WorkerGlobalScope* globalScope =
-      aWorkerPrivate->GetOrCreateGlobalScope(jsapi.cx());
-  if (NS_WARN_IF(!globalScope)) {
-    NS_WARNING("Failed to make global!");
-    // There's no way to report the exception on jsapi right now, because there
-    // is no way to even enter a compartment on this thread anymore.  Just clear
-    // the exception.  We'll report some sort of error to our caller in
-    // ShutdownScriptLoader, but it will get squelched for the same reason we're
-    // squelching here: all the error reporting machinery relies on being able
-    // to enter a compartment to report the error.
-    jsapi.ClearException();
-    return false;
-  }
-
   return true;
 }
 
@@ -2133,7 +2116,9 @@
     // Client execution ready and possible controlled by a service worker.
     if (mIsWorkerScript) {
       if (mScriptLoader.mController.isSome()) {
-        aWorkerPrivate->Control(mScriptLoader.mController.ref());
+        MOZ_ASSERT(mScriptLoader.mWorkerScriptType == WorkerScript,
+                   "Debugger clients can't be controlled.");
+        aWorkerPrivate->GlobalScope()->Control(mScriptLoader.mController.ref());
       }
       aWorkerPrivate->ExecutionReady();
     }
@@ -2294,8 +2279,13 @@
   Maybe<ClientInfo> clientInfo;
   Maybe<ServiceWorkerDescriptor> controller;
   if (!aIsMainScript) {
-    clientInfo = aWorkerPrivate->GetClientInfo();
-    controller = aWorkerPrivate->GetController();
+    nsIGlobalObject* global =
+        aWorkerScriptType == WorkerScript
+            ? static_cast<nsIGlobalObject*>(aWorkerPrivate->GlobalScope())
+            : aWorkerPrivate->DebuggerGlobalScope();
+
+    clientInfo = global->GetClientInfo();
+    controller = global->GetController();
   }
 
   RefPtr<ScriptLoaderRunnable> loader = new ScriptLoaderRunnable(
@@ -2428,7 +2418,12 @@
 
   // We are loading the main script, so the worker's Client must be
   // reserved.
-  info->mReservedClientInfo = aWorkerPrivate->GetClientInfo();
+  if (aWorkerScriptType == WorkerScript) {
+    info->mReservedClientInfo = aWorkerPrivate->GlobalScope()->GetClientInfo();
+  } else {
+    info->mReservedClientInfo =
+        aWorkerPrivate->DebuggerGlobalScope()->GetClientInfo();
+  }
 
   LoadAllScripts(aWorkerPrivate, std::move(aOriginStack), loadInfos, true,
                  aWorkerScriptType, aRv);